src/include/mutex_inline.h

```diff
@@ -266,6 +292,40 @@ __wt_spin_unlock(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
 
 #endif
 
+/*
+ * __wt_spin_owns --
+ *     Check whether the session owns the spinlock.
+ */
+static inline bool
+__wt_spin_owns(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
+{
+    return (t->session_id == WT_SPIN_SESSION_ID_SAFE(session));
+}
+
+/*
+ * __wt_spin_unlock_if_owns --
+ *     Unlock the spinlock only if it is acquired by the specified session.
+ */
+static inline void
+__wt_spin_unlock_if_owns(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
+{
+    if (__wt_spin_owns(session, t))
+        __wt_spin_unlock(session, t);
+}
+
+/*
+ * WT_ASSERT_SPINLOCK --
+ *      Assert that the session owns the spinlock.
+ */
+#define WT_ASSERT_SPINLOCK(session, t) WT_ASSERT((session), __wt_spin_owns((session), (t)));
+
+/*
+ * WT_ASSERT_SPINLOCK_COND --
+ *      Assert that the session owns the spinlock, but only if the given condition is satisfied.
+ */
+#define WT_ASSERT_SPINLOCK_COND(session, t, condition) \
```

Comment:
I would personally rather see the WT_ASSERT statement in the code rather than wrapped in this macro. It reads more explicitly to me, and feels clearer for the conditional one. It also allows for flexibility for future work that might want to use one of the other WT_ASSERT macros like WT_ASSERT_ALWAYS or WT_ASSERT_OPTIONAL without having to clutter up a lot more wrapper macros.

Reply:
To clarify, is it okay to keep the WT_ASSERT_SPINLOCK macro, even as we remove this _COND version? I'd really like to keep WT_ASSERT_SPINLOCK as it makes the assert a lot less verbose.

If we remove the _COND version, would it be okay to have something like this?
```
if (!F_ISSET(S2BT(session), WT_BTREE_SALVAGE))
    WT_ASSERT_SPINLOCK(session, &block->live_lock);
```
This would be probably the most readable way to do this. And I'd like to believe that a good compiler would figure out that it can remove even the `if` check in the release mode, because it has no side effect.

I originally had this with just WT_ASSERT calling the _owns function directly, but I found the condition in it not particularly readable, so I made this one.